### PR TITLE
SDKS-2543: Avoiding cache when update push notification arrives

### PR DESCRIPTION
**Pull request by javrudsky**
# Android SDK

## What did you accomplish?
- Adding No-Cache header when fetching changes when update or kill push notification received
- Added E2E and unit tests related
- Updated existing tests

### CODE REVIEW

File: src/main/java/io/split/android/client/network/HttpRequestImpl.java [hunk 1]
```diff
@@ -15,6 +15,7 @@
 import java.util.HashMap;
 import java.util.Map;
 
+import io.split.android.client.utils.Logger;
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/network/HttpRequestImpl.java:18]: unused import ?

File: src/main/java/io/split/android/client/service/mysegments/MySegmentsSyncTask.java [hunk 2]
```diff
@@ -69,10 +77,18 @@ private void logError(String message) {
...
+            return SplitHttpHeadersBuilder.noCacheHeaders();
+        }
+        return null;
+    }
+
     private void fireMySegmentsUpdatedIfNeeded(List<String> oldSegments, List<String> newSegments) {
         if(mEventsManager == null) {
             return;
         }
+        SplitInternalEvent e = getInternalEvent(oldSegments, newSegments);
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/service/mysegments/MySegmentsSyncTask.java:91]: `e` is not being used. should it be removed or some logic is missing?

File: src/main/java/io/split/android/client/service/sseclient/reactor/MySegmentsUpdateWorker.java [hunk 3]
```diff
@@ -31,7 +31,7 @@ public MySegmentsUpdateWorker(
     protected void onWaitForNotificationLoop() throws InterruptedException {
         try {
             mNotificationsQueue.take();
-            mSynchronizer.synchronizeMySegments();
+            mSynchronizer.forceMySegmentsSync();
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/service/sseclient/reactor/MySegmentsUpdateWorker.java:34]: check Javi, but `SynchronizerImpl::synchronizeMySegments` and `SynchronizerImpl::forceMySegmentsSync` seems to be doing the exact same thing. Consider removing one of them if that is the case, or probably `forceMySegmentsSync` requires to pass true for `avoidCache` flag.

File: src/main/java/io/split/android/client/service/synchronizer/SynchronizerImpl.java [hunk 4]
```diff
@@ -123,6 +121,13 @@ public void synchronizeSplits() {
 
     @Override
     public void synchronizeMySegments() {
+        mMySegmentsSyncRetryTimer.setTask(mSplitTaskFactory.createMySegmentsSyncTask(false), null);
+        mMySegmentsSyncRetryTimer.start();
+    }
+
+    @Override
+    public void forceMySegmentsSync() {
+        mMySegmentsSyncRetryTimer.setTask(mSplitTaskFactory.createMySegmentsSyncTask(false), null);
```
Comment (EmilianoSanchez) [inline src/main/java/io/split/android/client/service/synchronizer/SynchronizerImpl.java:130]: do you mean `false` --> `true` here?

**Review by EmilianoSanchez — CHANGES_REQUESTED** [review]

File: src/main/java/io/split/android/client/network/HttpRequestImpl.java (hunk 1 above)
Comment (javrudsky) [reply src/main/java/io/split/android/client/network/HttpRequestImpl.java:18]: Removed!

File: src/main/java/io/split/android/client/service/mysegments/MySegmentsSyncTask.java (hunk 2 above)
Comment (javrudsky) [reply src/main/java/io/split/android/client/service/mysegments/MySegmentsSyncTask.java:91]: Removed!

File: src/main/java/io/split/android/client/service/sseclient/reactor/MySegmentsUpdateWorker.java (hunk 3 above)
Comment (javrudsky) [reply src/main/java/io/split/android/client/service/sseclient/reactor/MySegmentsUpdateWorker.java:34]: Good catch Doc. Fixed!

File: src/main/java/io/split/android/client/service/synchronizer/SynchronizerImpl.java (hunk 4 above)
Comment (javrudsky) [reply src/main/java/io/split/android/client/service/synchronizer/SynchronizerImpl.java:130]: Fixed!

**Review by EmilianoSanchez — APPROVED** [review]
LGTM

**Pull request merged**